### PR TITLE
Adds special case to include transitive origins in GetInstalledAndTransitivePackagesAsync API

### LINKED ISSUE #12102: [Bug]: The transitive package doesn’t show in “Installed” tab until reopening the solution
### NuGet Product Used

Visual Studio Package Management UI

### Product Version

NuGet Client Dev\6.4.0.92

### Worked before?

_No response_

### Impact

It bothers me. A fix would be nice

## Repro steps: 
1.  Set the environment Variable NUGET_TRANSITIVE_DEPENDENCIES_IN_PMUI=1.
2. Create a new C# Console App (.Net Core) project.
3. Right-click the project in Solution Explorer and select ‘Manage NuGet Packages…’ menu item to open PM UI. 
 4. Click “Browse” tab, search a package (e.g. “Microsoft.EntityFrameworkCore”) that has dependencies, select a package version [6.0.4] in the "Version" drop-down list in the right panel and click the “Install” button on the right.
5. Switch to “Installed” tab.

## Expected:
The transitive package should show in “Installed” tab.

## Actual:
The transitive package doesn’t show in “Installed” tab as below screenshot.
image

## Notes: 
1. The transitive package can be show in “Installed” tab after closing solution/VS and reopening the same solution.
2. The issue starts from NuGet Client Dev\6.4.0.90.

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12102

Regression? Last working version: Yes, 17.3

## Description

1. Adds a special case to GetInstalledAndTransitivePackagesAsync (gitpa) API to calculate transitive origins:
    * If we have installed packages, transitive packages and no transitive origins data, and the gitpa API is _requesting_ transitive origins, then, the API will calculate transitive origins data and cache the results
2. Relaxes a condition to match package-typed dependencies from `PackageSpec` objects. This way, we are ensuring we will show installed packages found in assets file.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: 6 new unit tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A: No product changes.

### Validation

Animations provided below to compare the effects of this PR.

#### Before

The bug appears during package installation in PM UI: transitive packages are not shown after installing a package with transitive dependencies.

before_fix

#### After

Transitive packages appear after installing a package in PM UI.

after_fix

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/ProjectView.xaml.cs [hunk 1]
```diff
@@ -276,9 +276,7 @@ private void InstallButton_Clicked(object sender, RoutedEventArgs e)
 
         private void Versions_SelectionChanged(object sender, SelectionChangedEventArgs e)
         {
-            PackageDetailControlModel.PreviousSelectedVersion = e.AddedItems.Count > 0 ? e.AddedItems[0].ToString() : string.Empty;
-
-            return;
+            PackageDetailControlModel.PreviousSelectedVersion = e.AddedItems.Count > 0 ? e.AddedItems[0]?.ToString() : string.Empty;
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/ProjectView.xaml.cs]: This is not related to Transitve Dependencies fix, but it fixes an error for null references on telemetry.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs [hunk 2]
```diff
@@ -162,7 +162,9 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
             {
                 // Compute Transitive Origins
                 Dictionary<string, TransitiveEntry> transitiveOrigins;
-                if (IsInstalledAndTransitiveComputationNeeded || TransitiveOriginsCache == null)
+                if (IsInstalledAndTransitiveComputationNeeded // Cache invalidation
+                    || TransitiveOriginsCache == null // If any data race left the cache as null
+                    || (!TransitiveOriginsCache.Any() && calculatedTransitivePackages.Any() && includeTransitiveOrigins)) // We have transitive pacakges, but no transitive origins and the call is requesting transitive origins
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs]:
```suggestion
                    || (!TransitiveOriginsCache.Any() && calculatedTransitivePackages.Any() && includeTransitiveOrigins)) // We have transitive packages, but no transitive origins and the call is requesting transitive origins
```

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs]: It looks we're already checking `includeTransitiveOrigins` is true condition at line 161, so not sure if we need to check again.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs [hunk 3]
```diff
@@ -3915,12 +3904,172 @@ private static async Task<IPackageReferenceProject> PrepareTestProjectAsync(Simp
...
+        [Fact]
+        public async Task GetInstalledAndTransitivePackagesAsync_AlternateCallsToExcludeAndIncludeTransitiveOriginsInParallel_ReturnsCorrectDataAsync()
+        {
+            using var rootDir = new SimpleTestPathContext();
+
+            // Arrange
+            await CreatePackagesAsync(rootDir);
+            IPackageReferenceProject project = await PrepareTestProjectAsync(rootDir);
+            List<Task> tasks = new List<Task>();
+            for (int i = 0; i < 20; i++)
```
Comment (erdembayar) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs]: Isn't it too many? No sure how long this test takes.

**Review by erdembayar — COMMENTED** [review]
LGTM, just 2-3 comments.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs (hunk 3 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs]: It takes around 4 seconds to run the test on my local laptop. Lowering to 5 iterations, similar to other tests.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs [hunk 4]
```diff
@@ -162,7 +162,9 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
             {
                 // Compute Transitive Origins
                 Dictionary<string, TransitiveEntry> transitiveOrigins;
-                if (IsInstalledAndTransitiveComputationNeeded || TransitiveOriginsCache == null)
+                if (IsInstalledAndTransitiveComputationNeeded // Cache invalidation
+                    || TransitiveOriginsCache == null // If any data race left the cache as null
+                    || (!TransitiveOriginsCache.Any() && calculatedTransitivePackages.Any() && includeTransitiveOrigins)) // We have transitive packages, but no transitive origins and the call is requesting transitive origins
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs]: It looks like you left 1 more my comment: 
It looks we're already checking `includeTransitiveOrigins` is true condition at line 161, so not sure if we need to check again.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs]: Oh right! Thank you for pointing out this!

**Review by erdembayar — DISMISSED** [review]

**Review by nkolev92 — DISMISSED** [review]

**Review by erdembayar — APPROVED** [review]

**Pull request merged**